harness/determined/deploy/local/cluster_utils.py

```diff
@@ -177,56 +184,137 @@ def master_up(
 
-    # Ensure checkpoint storage directory exists.
-    final_storage_host_path = master_conf.get("checkpoint_storage", {}).get("host_path")
-    if final_storage_host_path is not None:
-        final_storage_host_path = Path(final_storage_host_path)
-        if not final_storage_host_path.exists():
-            final_storage_host_path.mkdir(parents=True)
-
     if make_temp_conf:
         fd, temp_path = tempfile.mkstemp(prefix="det-deploy-local-master-config-")
         with open(fd, "w") as f:
             yaml.dump(master_conf, f)
-        master_config_path = Path(temp_path)
+        master_config_path = pathlib.Path(temp_path)
 
     # This is always true by now, but mypy needs help.
     assert master_config_path is not None
+    restart_policy = "unless-stopped" if autorestart else "no"
+    env = {
+        "DET_MASTER_HTTP_PORT": str(port),
+        "DET_DB_PASSWORD": db_password,
+        "DET_LOG_INFO": "info",
+    }
 
-    env["DET_MASTER_CONFIG"] = str(master_config_path.resolve())
+    # Kill existing master container if exists.
+    master_down(master_name=master_name, delete_db=delete_db, cluster_name=cluster_name)
 
-    master_down(master_name, delete_db)
-    docker_compose(command, master_name, env)
-    _wait_for_master("localhost", port, cluster_name)
+    # Create network.
+    client = docker_client()
+    print(f"Creating network {NETWORK_NAME}...")
+    client.networks.create(name=NETWORK_NAME, attachable=True)
+
+    # Start db and wait for healthcheck.
+    db_name = f"{cluster_name}_{DB_NAME}"
+    db_up(name=db_name, password=db_password, network_name=NETWORK_NAME, cluster_name=cluster_name)
+    try:
+        _wait_for_container(db_name)
+    except TimeoutError:
+        print(f"Database {db_name} failed to reach a ready state, removing container.")
+        _kill_containers(names=[db_name])
+        raise TimeoutError
+
+    # Start master instance.
+    print(f"Creating {master_name}...")
+    volumes = [f"{os.path.abspath(master_config_path)}:/etc/determined/master.yaml"]
+    client.containers.run(
+        image=f"{image_repo_prefix}/determined-master:{version}",
+        environment=env,
+        init=True,
+        mounts=[],
+        volumes=volumes,
+        name=master_name,
+        detach=True,
+        labels={},
+        restart_policy={"Name": restart_policy},
+        device_requests=None,
+        ports={"8080": f"{port}"},
+    )
 
+    # Connect to the network separately to set alias.
+    network = client.networks.get(NETWORK_NAME)
+    network.connect(container=master_name, aliases=["determined-master"])
 
-def master_down(master_name: str, delete_db: bool) -> None:
+    try:
+        _wait_for_master("localhost", port, cluster_name)
+    except TimeoutError:
+        print(f"Master failed to initialize, removing master: {master_name} and DB {db_name}.")
+        master_down(master_name=master_name, delete_db=delete_db, cluster_name=cluster_name)
+        raise TimeoutError
```

Comment:
what about the db container?

Reply:
Side note, you might find a `contextlib.ExitStack` to be a more expressive way to handle this.

Kind of like my `defer` pattern in the `_pytorch_trial.py`.  You could defer your teardown code for network and database.  Then, the last thing you'd do before exiting the `ExitStack` context manager successfully would be an `exit_stack.pop_all()` to prevent the deferred cleanups from happening in success cases.

Reply:
beautiful.